### PR TITLE
fix: seperate nodes entry for backward compatibility

### PR DESCRIPTION
**Pull request by yue4u**
This PR attempts to address compatibility issues with three before r161.

Removed `export * from './nodes'` in `three-vrm-materials-mtoon` and created seperated `./nodes` entry in both `three-vrm-materials-mtoon` and `three-vrm`

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -3,7 +3,7 @@
     "target": "es6",
     "module": "es6",
     "lib": ["es2018", "dom"],
-    "moduleResolution": "node",
+    "moduleResolution": "Bundler",
```
Comment (yue4u) [inline tsconfig.json:6]: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-0.html#--moduleresolution-bundler
Expect this not to affect end user

File: packages/three-vrm/src/nodes/index.ts [hunk 2]
```diff
@@ -0,0 +1 @@
+export * from '@pixiv/three-vrm-materials-mtoon/nodes';
```
Comment (yue4u) [inline packages/three-vrm/src/nodes/index.ts:1]: Should we add some three version checks and log info here or inside `@pixiv/three-vrm-materials-mtoon/nodes`? IMO it's not necessary

File: README.md [hunk 3]
```diff
@@ -92,7 +92,7 @@ Code like this:
 
 ```javascript
 import * as THREE from 'three';
-import { GLTFLoader } from 'three/examples/jsm/loaders/GLTFLoader';
+import { GLTFLoader } from 'three/examples/jsm/loaders/GLTFLoader.js';
```
Comment (yue4u) [inline README.md:95]: align to three export

File: bin/build.mjs [hunk 4]
```diff
@@ -55,16 +55,46 @@ async function buildPackage(absWorkingDir) {
...
+    /**
+     * @param {string} base
+     */
+    const addSuffix = (base) => {
+      let outFilename = base + (format === 'esm' ? '.module' : '');
+      return outFilename + (DEV ? '' : '.min');
+    };
+    let outFilename = addSuffix(filename);
+
+    const extraEntryPoints = ['@pixiv/three-vrm', '@pixiv/three-vrm-materials-mtoon'].includes(packageJson.name)
```
Comment (yue4u) [inline bin/build.mjs]: doing this here to modify cjs,esm outputs altogether

File: bin/build.mjs [hunk 5]
```diff
@@ -55,16 +55,46 @@ async function buildPackage(absWorkingDir) {
...
     return {
       entryPoints: {
         [outFilename]: 'src/index.ts',
+        ...extraEntryPoints,
       },
       format,
     };
   };
 
+  /**
```
Comment (yue4u) [inline bin/build.mjs]: Coundn't find a better way to do this. IMO it's too wasteful to bundle shaders and everything from three-vrm-materials-mtoon main entry again

File: tsconfig.json (hunk 1 above)
Comment (0b5vr) [reply tsconfig.json:6]: why is the first letter capitalized though?

File: packages/three-vrm/src/nodes/index.ts (hunk 2 above)
Comment (0b5vr) [reply packages/three-vrm/src/nodes/index.ts:1]: we already have a warning inside `MToonNodeMaterial` so I think it's not needed

File: tsconfig.json (hunk 1 above)
Comment (yue4u) [reply tsconfig.json:6]: From autocompletion. (guess this is the default)
Screenshot 2024-06-28 at 13 32 36

File: packages/three-vrm/src/nodes/index.ts (hunk 2 above)
Comment (yue4u) [reply packages/three-vrm/src/nodes/index.ts:1]: I think MToonNodeMaterial'constructor will never be executed if `three/addons/nodes/Nodes` not exist?

File: bin/build.mjs (hunk 4 above)
Comment (0b5vr) [reply bin/build.mjs]: This section looks surprisingly cryptic compared to what it actually does 😩

File: bin/build.mjs [hunk 6]
```diff
@@ -55,16 +55,46 @@ async function buildPackage(absWorkingDir) {
...
+     */
+    const addSuffix = (base) => {
+      let outFilename = base + (format === 'esm' ? '.module' : '');
+      return outFilename + (DEV ? '' : '.min');
+    };
+    let outFilename = addSuffix(filename);
+
+    const extraEntryPoints = ['@pixiv/three-vrm', '@pixiv/three-vrm-materials-mtoon'].includes(packageJson.name)
+      ? { [addSuffix('nodes/index')]: 'src/nodes/index.ts' }
+      : {};
```
Comment (0b5vr) [inline bin/build.mjs]:
```suggestion
    const needsNodesBuild = ['@pixiv/three-vrm', '@pixiv/three-vrm-materials-mtoon'].includes(packageJson.name);
    const extraEntryPoints = needsNodesBuild
      ? { [addSuffix('nodes/index')]: 'src/nodes/index.ts' }
      : {};
```

File: bin/build.mjs [hunk 7]
```diff
@@ -55,16 +55,46 @@ async function buildPackage(absWorkingDir) {
    * @param {'esm' | 'cjs'} format
    */
   const entryPoints = (format) => {
-    let outFilename = filename + (format === 'esm' ? '.module' : '');
-    outFilename += DEV ? '' : '.min';
+    /**
+     * @param {string} base
+     */
```
Comment (0b5vr) [inline bin/build.mjs]:
```suggestion
    /**
     * Add suffix according to the format to the given basename of the output file.
     * @param {string} base Basename of the output file
     * @returns {string} `<base>[.module][.min]`
     */
```

File: bin/build.mjs [hunk 8]
```diff
@@ -55,16 +55,46 @@ async function buildPackage(absWorkingDir) {
...
+    };
+    let outFilename = addSuffix(filename);
+
+    const extraEntryPoints = ['@pixiv/three-vrm', '@pixiv/three-vrm-materials-mtoon'].includes(packageJson.name)
+      ? { [addSuffix('nodes/index')]: 'src/nodes/index.ts' }
+      : {};
+
     return {
       entryPoints: {
         [outFilename]: 'src/index.ts',
```
Comment (0b5vr) [inline bin/build.mjs]: I think we can directly put `addSuffix(filename)` here as `extraEntryPoints` does

File: bin/build.mjs (hunk 5 above)
Comment (0b5vr) [reply bin/build.mjs]: I will find a way to quit `MToonNodeMaterialLoaderPlugin`. Maybe having a factory pattern or something would help this

File: tsconfig.json (hunk 1 above)
Comment (0b5vr) [reply tsconfig.json:6]: huh. It's kinda weird for me but it's ok.

File: bin/build.mjs (hunk 5 above)
Comment (yue4u) [reply bin/build.mjs]: removed in 6d40b26a5d52d82c00d40b66f72874e5c507744b

File: bin/build.mjs (hunk 8 above)
Comment (yue4u) [reply bin/build.mjs]: [e2e96c9](https://github.com/pixiv/three-vrm/pull/1427/commits/e2e96c9a27992cf7001ba81e27350643ae97dc7f)

**Comment by yue4u** [comment]
webgpu-dnd example now produces a lot of errors when using with an order version of three. Can we do something about this?

Screenshot 2024-07-01 at 15 35 05

**Comment by 0b5vr** [comment]
> webgpu-dnd example now produces a lot of errors when using with an order version of three. Can we do something about this?

`WebGPURenderer.renderAsync` has been introduced in r161. I believe it's fine in all versions of r161 or later

See: https://github.com/mrdoob/three.js/pull/27597

File: packages/three-vrm/src/nodes/index.ts (hunk 2 above)
Comment (0b5vr) [reply packages/three-vrm/src/nodes/index.ts:1]: Ahh, yes it's possible to cause an error before `MToonNodeMaterial` instantiates.

Does this help?

image

**Review by 0b5vr — APPROVED** [review]

**Pull request merged**